pandas/core/case_when.py

```diff
@@ -0,0 +1,167 @@
+        If this value is not already an array like (i.e. it is not of type `Series`,
+        `np.array` or `list`) it will be repeated `obj.shape[0]` times in order to
+        create an array like object from it and then apply the `Series.mask`.
+
+    Returns
+    -------
+    Series
+        Series with the corresponding values based on the conditions, their values
+        and the default value.
+
+
+    Examples
+    --------
+    >>> df = pd.DataFrame(dict(a=[1, 2, 3], b=[4, 5, 6]))
+    >>> df
+       a  b
+    0  1  4
+    1  2  5
+    2  3  6
+
+    >>> pd.case_when(
+    ...     df,
+    ...     lambda x: x.a == 1,
+    ...     'first',
+    ...     lambda x: (x.a == 2) & (x.b == 5),
+    ...     'second',
+    ...     default='default',
+    ... )
+    0      first
+    1     second
+    2    default
+    dtype: object
+
+    >>> pd.case_when(
+    ...     df,
+    ...     lambda x: (x.a == 1) & (x.b == 4),
+    ...     df.b,
+    ...     default=0,
+    ... )
+    0    4
+    1    0
+    2    0
+    dtype: int64
+
+    >>> pd.case_when(
+    ...     df,
+    ...     lambda x: (x.a > 1) & (x.b > 1),
+    ...     -1,
+    ...     default=df.a,
+    ... )
+    0    1
+    1   -1
+    2   -1
+    Name: a, dtype: int64
+
+    >>> pd.case_when(
+    ...     df.a,
+    ...     lambda x: x == 1,
+    ...     -1,
+    ...     default=df.a,
+    ... )
+    0   -1
+    1    2
+    2    3
+    Name: a, dtype: int64
+
+    >>> pd.case_when(
+    ...     df.a,
+    ...     df.a  > 1,
+    ...     -1,
+    ...     default=df.a,
+    ... )
+    0    1
+    1   -1
+    2   -1
+    Name: a, dtype: int64
+
+    The index will always follow that of `obj`. For example:
+    >>> df = pd.DataFrame(
+    ...     dict(a=[1, 2, 3], b=[4, 5, 6]),
+    ...     index=['index 1', 'index 2', 'index 3']
+    ... )
+    >>> df
+             a  b
+    index 1  1  4
+    index 2  2  5
+    index 3  3  6
+
+    >>> pd.case_when(
+    ...     df,
+    ...     lambda x: (x.a == 1) & (x.b == 4),
+    ...     df.b,
+    ...     default=0,
+    ... )
+    index 1    4
+    index 2    0
+    index 3    0
+    dtype: int64
+    """
+    len_args = len(args)
+
+    if len_args < 2:
+        raise ValueError("At least two arguments are required for `case_when`")
+    if len_args % 2:
+        raise ValueError(
+            "The number of conditions and values do not match. "
+            f"There are {len_args - len_args//2} conditions "
+            f"and {len_args//2} values."
+        )
+
+    # construct series on which we will apply `Series.mask`
+    series = pd.Series(default, index=obj.index)
+
+    for i in range(0, len_args, 2):
+        # get conditions
+        if callable(args[i]):
+            conditions = com.apply_if_callable(args[i], obj)
+        else:
+            conditions = args[i]
+
+        # get replacements
+        replacements = args[i + 1]
+
+        # `Series.mask` call
+        series = series.mask(conditions, replacements)
```

Comment:
I believe we want to only modify the value on the first `conditions` that evaluates to True (correct me if you think this is wrong!); this will modify it for all conditions. One approach is to maintain a `modified` that starts as `pd.Series(False, index=self.index)`. Then this line can become `series = series.mask(~modified & conditions, replacements)`. After this line, we also need to update `modified |= conditions`.

Reply:
Good point. Updated ([f2bf4d8](https://github.com/pandas-dev/pandas/pull/50343/commits/f2bf4d866a38fd888561b4c3d2acc537051b7cdc)). 

Now, if multiple conditions are met, the value of the first one is used. For example: 

```
>>> df 
         a  b
index 1  1  4
index 2  2  5
index 3  3  6

>>> pd.case_when(
... df,
... lambda x: x.a > 0,
... 1,
... lambda x: x.a == 1, 
... -1,
... default='default'
)
index 1    1
index 2    1
index 3    1
dtype: object